### PR TITLE
feat: ACNA-1602 - add logic for --no-action-code and add options object as a param 

### PR DESCRIPTION
**Pull request by arjuncooliitr**
## Description

To enable preventing action code deployment using `--no-action-code `flag :

- Added options objects as a param to relevant methods
- Conditional logic to prevent adding action code on method `createActionObject`

## Dependents

- https://github.com/adobe/aio-cli-plugin-app/issues/537

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by moritzraho — APPROVED** [review]
lgtm - just one nitpick

File: src/deploy-actions.js
Comment (arjuncooliitr) [reply src/deploy-actions.js]: Removed the word 'flag' from the `options` description to stay consistent with this module as suggested by @moritzraho. 
Also, changed the param name from `noCode` to `actionCode` to simplify and avoid using negated version of the param.

File: src/deploy-actions.js
Comment (shazron) [inline src/deploy-actions.js:41]: For the tests:
1. Change `deployActions` signature to:
`async function deployActions (config, deployConfig = {}, log = console.log, options = {}) {`
2. Remove line https://github.com/adobe/aio-lib-runtime/blob/0e6ce80979816bd2ed1712ceed849702f164d38c/src/deploy-actions.js#L43

This should at least have defaults for all the parameters, but it will not complete coverage.

**Review by shazron — CHANGES_REQUESTED** [review]

**Comment by arjuncooliitr** [comment]
Unit Tests (UTs)were failing with mainly 2 types of errors:
1. UTs for `createActionObject` -`actionCode` is undefined property:  I added default to options param -`options = { actionCode: true }` , which resolved these errors. Also, it kind of made sense to me since, by default we want to include action code unless we unset the flag.
2. UTs for `deployActions` : They should now expect an additional parameter `options`, so modified the UTs accordingly.

All the units tests are now passing although the coverage is still not 100%. Looking to add new UTs for that.
Please review these changes @shazron

**Review by shazron — CHANGES_REQUESTED** [review]

File: src/deploy-actions.js
Comment (arjuncooliitr) [reply src/deploy-actions.js:41]: We are calling `deployActions` from app `deploy`command, where `options` could not be an empty object since we are creating it just above using it : `const options = { actionCode: flags['action-code'] }`
So, I am not sure if this case would arise: `deployActions (config, {}, null, {})`. Please correct me if I am wrong. @shazron

**Comment by arjuncooliitr** [comment]
Not able to get 100% branch coverage for `deploy-action.js`. Currently its 99.85%.

Line 148 - `async function deployWsk (scriptConfig, manifestContent, logFunc, filterEntities, options = { actionCode: true }) {}` is missing coverage but not sure why. It is not a conditional statement so what branches are being missed? Need a bit help here @shazron @moritzraho

**Comment by arjuncooliitr** [comment]
Actually after removing default value for options param, I am able to get 100% branch coverage
`async function deployWsk (scriptConfig, manifestContent, logFunc, filterEntities, options) {}`
Trying understand this behaviour

image

**Comment by shazron** [comment]
@arjuncooliitr after running the test locally, check the `coverage` folder on your repo. There should be an `index.html` file that you can load that will give you detailed coverage

File: src/deploy-actions.js
Comment (arjuncooliitr) [reply src/deploy-actions.js:41]: @moritzraho Since Shaz is not here for the week to review this conversation, what do you think on it?

Comment (moritzraho) [reply src/deploy-actions.js:41]: I +1 Shaz's suggestion to set `actionCode = true` default option

Comment (arjuncooliitr) [reply src/deploy-actions.js]: Thanks @moritzraho for the comments. I made suggested changes to the default param in the APIs. Can you please review these changes once and approve the PR if everything seems alright.

File: src/deploy-actions.js
Comment (arjuncooliitr) [reply src/deploy-actions.js]: Removed default value of param from the function signature

File: src/utils.js
Comment (arjuncooliitr) [reply src/utils.js]: Made suggested changes

File: src/utils.js
Comment (arjuncooliitr) [reply src/utils.js]: Made suggested changes

File: src/deploy-actions.js
Comment (moritzraho) [inline src/deploy-actions.js:59]: nitpick, `optionsParam = options`  will allow modifications on the `options` input object downstream, which might be a source for bugs in the future. maybe do something like `const optionsParm = { actionCode: options.actionCode !== false }` instead

File: src/utils.js
Comment (moritzraho) [inline src/utils.js:1194]: same here

File: src/deploy-actions.js
Comment (arjuncooliitr) [reply src/deploy-actions.js]: Sure @moritzraho . Thanks for pointing that out. I have made `optionsParam` as a `const`

**Review by moritzraho — APPROVED** [review]
lgtm thanks !

File: src/utils.js
Comment (sandeep-paliwal) [inline src/utils.js:992]: Instead of this check we can add actionCode = true as default value for options param

Comment (arjuncooliitr) [reply src/utils.js:992]: Hi @sandeep-paliwal . So good to have you back! :)
Thanks for the review. Initially, I added actionCode = true as the default value for options param but then Shaz rightly pointed out that there could be a case when the API is called with empty options object:
`createActionObject (fullName, manifestAction,{})`
In that case default value wouldn't apply. 
Link to the above conversation: https://github.com/adobe/aio-lib-runtime/pull/98#discussion_r891558501

**Review by shazron — APPROVED** [review]

**Comment by shazron** [comment]
Long overdue - if no more comments I will merge at the end of the week.

**Comment by shazron** [comment]
Need to resolve concerns in https://github.com/adobe/aio-cli-plugin-app/pull/537 first.

**Comment by purplecabbage** [comment]
This is being closed in favor of updated documentation.  This pull request was well implemented, however the convenience was deemed to potentially lead to developers overwriting actions as the command syntax was awkward.  
The command `aio runtime action update --param` is expected to be used instead of surfacing the command as an additional flag on `aio app deploy`

**Pull request closed without merging**
